### PR TITLE
electron: fix backend forking

### PR DESCRIPTION
**Pull request by marechal-p**
#### What it does

This PR fixes 3 issues:

1. When forking, Electron 4 forgets to set `process.versions.electron` in
the sub-process, making some packages wrongly guess their environment.
This commit manually patches this variable.

2. The generated application code also used to not fork the backend when
working from sources. Running the backend in the Electron main process
was only meant to help with debugging. This commit makes the generated
code fork the backend unless a `--no-cluster` flag is specified.

3. When running a bundled application, the backend was forked. But any
command line parameter passed to the Electron application wasn't passed
down to the backend. This commit makes sure that relevant arguments
are passed to the forked backend process.
\
If you want the generated code to not fork the backend in a sub-process,
you can either pass a `--no-cluster` flag to your application, or set
`THEIA_ELECTRON_NO_BACKEND_FORK=1` environment variable.

Closes: https://github.com/eclipse-theia/theia/pull/7361
Fixes: #7358
Fixes: #4288
Fixes: #4875

#### How to test

Add `grpc` as a dependency, and add `require('grpc')` in `examples/electron/src-gen/backend/server.js`. If you place the require before the patching of the `process.versions.electron` it should not work, but it should work if you put the require after.

note: `grpc` might need to be configured to run with Electron. What I did was:

```sh
cd node_modules/grpc
yarn node-pre-gyp --runtime=electron --target=4.2.12 install
```

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

**Comment by marechal-p** [comment]
Most changes are just related to debugging, as I tried to make sure that we could still debug code in both processes (electron main + node).

The fork-related changes are the following:
- https://github.com/eclipse-theia/theia/pull/7386/files#diff-473120a6fb850733ee88d073aac2b52cR31-R34
- https://github.com/eclipse-theia/theia/pull/7386/files#diff-e6697ec7c0f7ddbbfd9e8365baacf132R341-R363

**Comment by akosyakov** [comment]
Why don't keep dev mode and make a minimal change?

**Comment by marechal-p** [comment]
@akosyakov It will make Electron APIs fail right away in the backend when working in our repo. Before, we used to be able to require electron, but it would fail once packaged.

**Comment by akosyakov** [comment]
I think it only makes debugging harder and does not make it close to production. You have to bundle to be closer to production.  I think PR should be simpler and don't try to solve unrelated issues.

**Comment by marechal-p** [comment]
I stand by the fact that it will make Electron APIs fail on the backend this way.

**Comment by akosyakov** [comment]
> I stand by the fact that it will make Electron APIs fail on the backend this way.

It was not a problem so far, was it?

**Comment by marechal-p** [comment]
And now it won't ever!

But if you really feel strongly about separating electron main process and backend process making it too hard for developers to work with, I can make the PR even more minimal, I just don't think we should strive for absolutely minimal changes, but rather reasonable changes.

Please confirm that you really want to keep the old logic, for me to be sure.

**Comment by akosyakov** [comment]
I mean that all products and extensions are relying on it for debugging Electron right now: https://github.com/theia-ide/generator-theia-extension/blob/master/templates/launch.json#L30-L58

Including docs in This repo: https://github.com/eclipse-theia/theia/blob/master/doc/Developing.md#debug-the-electron-examples-backend

That's for sure not a trivial change without consequences.

**Comment by marechal-p** [comment]
The internal documentation can be updated, the generator is annoying indeed but it doesn't feel complete either. If anything, this makes me think that the generator package should be inside the theia repository itself, so that it is easier to align with changes we make.

If we want to also hide implementation details regarding debugging from client POV, we could also make a VS Code extension contributing a `theia` debugger type. WDYT?

I will bring back old behavior based on what you pointed out.

**Comment by akosyakov** [comment]
> If anything, this makes me think that the generator package should be inside the theia repository itself, so that it is easier to align with changes we make.

It is not about generator. It is that debugging setup in the Arduino Pro IDE won't work anymore and all other products out there, including 3rd party Theia extensions, workshops materials and so on.

> I will bring back old behavior based on what you pointed out.

Thank you.

**Comment by akosyakov** [comment]
But moving the generator sounds like a good idea. Could you file a proposal issue to see whether it is alright with everybody? I was thinking to change it to theia generator instead of extension, so it can generate different layouts, i.e. to support Theia extension development, VS Code extension development against Theia or a hybrid approach. But it should be another issue and discussion.

**Comment by marechal-p** [comment]
Updated the PR.

**Comment by akosyakov** [comment]
@marechal-p thank you Paul!

 May I ask you to add comments to
- `devMode` that we cannot remove it, since end products rely on it with a link to the discussions in this PR
- forking that we cannot remove it because running the express server in the electron main process can compromise UI performance with a link to the discussion in @kittaakos PR

I'm pretty sure that it was discussed before It would be document it this time that we won't wonder about it next year again.

**Comment by akosyakov** [comment]
@kittaakos Could you have a look please? I agree with @marechal-p that running the express server in the electron main thread is not a good idea. I'm pretty sure we discussed it with @svenefftinge long time ago and decided to run in a separate process to avoid compromising performance of electron.

**Comment by kittaakos** [comment]
I have checked the changes, thank you! I would leave it as is, the `process.versions.electron` patching I can do before calling `electron-builder` too. No need to change anything in Theia.

**Comment by marechal-p** [comment]
@kittaakos What do you mean?

**Comment by marechal-p** [comment]
> We run in dev mode only for debugging, not running from the command line.

@akosyakov Did not see this bit earlier, but we do run in "devMode" when running from sources, not just when debugging.

**Comment by marechal-p** [comment]
@akosyakov Added comments.

**Comment by akosyakov** [comment]
@marechal-p I thought it is based on [--no-cluster](https://github.com/eclipse-theia/theia/blob/cc9453d3a01644f371e9e94134b93ff65e1e4112/.vscode/launch.json#L33) option meaning that electron main and express server should run in the same process. If it is based on something else you will need to dig why it was introduced and whether it is still relevant and whether it can be changed without affecting setups of downstream projects.

**Comment by marechal-p** [comment]
> If it is based on something else you will need to dig why it was introduced and whether it is still relevant and whether it can be changed without affecting setups of downstream projects.

If I remove it, 2 processes will be started when running from sources, leading to debugging issues like you mentioned before. @thegecko has a requirement that might make us reconsider spawning that process or not btw, see https://github.com/eclipse-theia/theia/pull/7361#issuecomment-602676877

**Comment by akosyakov** [comment]
> If I remove it, 2 processes will be started when running from sources, leading to debugging issues like you mentioned before.

I think if you still respect `--no-cluster` option it should be fine.

**Comment by marechal-p** [comment]
@akosyakov Made the Electron generated code handle this `--no-cluster` flag.

But this made me see that arguments were not passed to the forked backend. I was able to confirm using theia-apps/theia-electron in a bundled application that the backend could not see any argument passed from the command line. This commit fixes it, although I could make it a second commit.

**Comment by marechal-p** [comment]
Broken the PR in smaller commits.

File: dev-packages/application-manager/README.md [hunk 1]
```diff
@@ -10,6 +10,23 @@
...
+This package contains the generators used when creating a Theia application through `@theia/cli`.
+
+This package can auto-generate application code for both the backend and frontend, as well as webpack configuration files.
+
+When targeting Electron, the `electron-main.js` script will spawn the backend process in a Node.js sub-process, where Electron's API won't be available.
+To prevent the generated application from forking the backend, you can pass a `--no-cluster` flag, or set an environment variable like `THEIA_ELECTRON_NO_BACKEND_FORK=1`.
+
+```sh
+# when developing a Theia application with @theia/cli:
+yarn theia start --no-cluster
```
Comment (akosyakov) [inline dev-packages/application-manager/README.md]: `theia` cli is documented here: https://github.com/eclipse-theia/theia/blob/master/dev-packages/cli/README.md#outline

Comment (marechal-p) [reply dev-packages/application-manager/README.md]: Fixed.

**Comment by akosyakov** [comment]
> https://github.com/eclipse-theia/theia/pull/7386#issuecomment-602591032

@kittaakos Could you elaborate why patching is not good? Does not it fix all downstreams or you think it is not a common problem?

@marechal-p maybe you move other changes to a new PR that they can be merged. I don't want to approve the patching because I am not so familiar with Electron products and Akos can know something  what we overlook.

**Comment by kittaakos** [comment]
> @kittaakos Could you elaborate why patching is not good?

I did not say, it is not good. It is a hack. I can do the same, in fact, I already have a workaround for the problem.

File: dev-packages/cli/README.md [hunk 2]
```diff
@@ -270,6 +270,20 @@ The property `theiaPlugins` describes the list of plugins to download, for examp
 }
 ```
 
+## Autogenerated Application
```
Comment (akosyakov) [inline dev-packages/cli/README.md]: it would be good to update TOC above

**Comment by akosyakov** [comment]
> I did not say, it is not good. It is a hack. I can do the same, in fact, I already have a workaround for the problem.

Ok, i don't understand whether it fixes the issue or not and whether it helps anyone. Also don't know how to test it. `How to test` section is empty. It is up to you whether this PR should be approved or not.

**Comment by marechal-p** [comment]
> I can do the same, in fact, I already have a workaround for the problem.

WDYM? What would people do that I should not do here?

> It is a hack.

It is a bug, that we manually fix. At worst, this is monkey patching.

But to argument in favor of doing this:

1. Electron 8.x.x automatically does it, it looks like a bug with 4.x.x that this variable is not set.
2. Forked Electron processes are actually Electron processes, they are not your regular Node.js runtime. What I mean is that you will need native libraries compiled for Electron to run inside it.
3. Most tools rely on this `process.versions.electron` to use Electron-specific features:
    - `is-electron` https://github.com/cheton/is-electron/blob/master/index.js#L8-L11
    - `node-pre-gyp` https://github.com/mapbox/node-pre-gyp/blob/674fda7b28b86bb3b11889db86c15c4d06722d02/lib/util/versioning.js#L262-L269

Tell me if you see anything else that would make this patching incomplete.

**Comment by marechal-p** [comment]
@akosyakov I have updated the `How to test` section, the use case comes from https://github.com/eclipse-theia/theia/issues/7358

File: dev-packages/application-manager/src/generator/frontend-generator.ts [hunk 3]
```diff
@@ -290,6 +290,8 @@ app.on('ready', () => {
     // Check whether we are in bundled application or development mode.
     // @ts-ignore
     const devMode = process.defaultApp || /node_modules[\/]electron[\/]/.test(process.execPath);
+    // Check if user wants to run everything as one process.
+    const noBackendFork = Boolean(process.env['THEIA_ELECTRON_NO_BACKEND_FORK']) || process.argv.includes('--no-cluster');
```
Comment (akosyakov) [inline dev-packages/application-manager/src/generator/frontend-generator.ts]: I'm not sure about adding `process.env['THEIA_ELECTRON_NO_BACKEND_FORK']`. I'm afraid to open up the approach which potentially breaks performance.

@thegecko Will you use it to run the server process inside the electron main?

File: dev-packages/cli/README.md [hunk 4]
```diff
@@ -270,6 +271,20 @@ The property `theiaPlugins` describes the list of plugins to download, for examp
...
+
+This package can auto-generate application code for both the backend and frontend, as well as webpack configuration files.
+
+When targeting Electron, the `electron-main.js` script will spawn the backend process in a Node.js sub-process, where Electron's API won't be available. To prevent the generated application from forking the backend, you can pass a `--no-cluster` flag, or set an environment variable like `THEIA_ELECTRON_NO_BACKEND_FORK=1`.
```
Comment (akosyakov) [inline dev-packages/cli/README.md]: We should be very clear that:
- `--no-cluster` should be never used in production, it runs everything in one process, like express server, file watching, git location and so on
- `THEIA_ELECTRON_NO_BACKEND_FORK` can cause performance issues by running the express server within the electron main process, with a link to https://www.electronjs.org/docs/tutorial/performance#3-blocking-the-main-process

**Comment by kittaakos** [comment]
> Tell me if you see anything else that would make this patching incomplete.

Thanks for preparing it, @marechal-p. I am going to verify your changes on a bundled electron app I work on and get back to you this week. Bear with me.

**Comment by marechal-p** [comment]
> For sure, I won't use `2.` and `3.` in the near future, but if you want this modification, I have no objections against them.

I do, thanks for the review!

File: dev-packages/cli/README.md (hunk 4 above)
Comment (marechal-p) [reply dev-packages/cli/README.md]: Removed the env var.

**Comment by marechal-p** [comment]
> @marechal-p, next time let's spend the same amount of effort we and especially you have spent on this issue with the electron 8 migration. Now, we have this `process.version.electron` fix, which has not caused a problem for anyone so far, except with `grpc`.

This was a quick fix, but I would be happy to help with Electron 8 once we get there!

**Pull request merged**